### PR TITLE
Enable contenthash in filename for dynamic assets

### PR DESCRIPTION
**Pull request by silverwind**
This should solve the main problem of dynamic assets caches getting stale after a version upgrade. Everything not affected will use querystring-based cache-busting, which includes files loaded via HTML or the worker scripts.

I tested with monaco and the PNG of jquery-minicolors and it's working fine for those.

This is basically a stripped-down version of https://github.com/go-gitea/gitea/pull/18632 which does not go all the way with contenthash in all filenames (as that requires non-trivial backend changes as well), but should be sufficient to solve the immediate issue of stale browser cache after upgrade.

## :warning: BREAKING :warning:

This PR removes the now unused `MD5` template function. Users with custom templates will need to update their templates to stop using this function.

### CODE REVIEW

File: webpack.config.js
Comment (silverwind) [inline webpack.config.js]:
```suggestion
      return `js/${language ? `monaco-language-${language.toLowerCase()}` : `[name]`}.[contenthash:8].js`;
```

**Comment by silverwind** [comment]
Did one improvement to service worker path comparison, as I have not much faith on whether [scriptUrl](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorker/scriptURL) is consistent across browsers in regards to query string or absolute/relative URL.

File: web_src/js/testUtils/setup.js
Comment (lafriks) [inline web_src/js/testUtils/setup.js]: It would be better to use `https://try.gitea.io`

Comment (silverwind) [inline web_src/js/testUtils/setup.js]:
```suggestion
window.location = new URL('https://try.gitea.io');
```

File: web_src/js/utils.test.js
Comment (silverwind) [inline web_src/js/utils.test.js:115]:
```suggestion
  expect(parseUrl('').pathname).toEqual('/');
  expect(parseUrl('/path').pathname).toEqual('/path');
```

File: web_src/js/utils.js
Comment (silverwind) [inline web_src/js/utils.js]: FYI, this should ideally use the `base` option of `URL()` instead of concatenating, but I [hit a bug in happy-dom](https://github.com/capricorn86/happy-dom/issues/569) which prevents it. Maybe we should switch back to jsdom as happy-dom seems immature in some aspects like this one.

Comment (silverwind) [reply web_src/js/utils.js]: I've done the switch to `jsdom`, no workarounds necessary with it anymore as it sets `window.location` to `http://localhost` and uses the reference `URL` implementation.

**Review by zeripath — APPROVED** [review]

**Review by lunny — APPROVED** [review]

File: templates/base/head_script.tmpl
Comment (wxiaoguang) [inline templates/base/head_script.tmpl]: I do not think here need another variable or MD5 hash (more CPU time).

`res.js?v=${appVer}` should do the work and is clear.

Comment (zeripath) [reply templates/base/head_script.tmpl]: agh! I missed this!

Agreed with @wxiaoguang here. 

This MD5 is being recalced every time this page is requested - this is a bad idea !

**Review by zeripath — CHANGES_REQUESTED** [review]
as per @wxiaoguang's comment

File: templates/base/head_script.tmpl
Comment (silverwind) [reply templates/base/head_script.tmpl]: `MD5 AppVer` is used in multiple places in HTML already for cache busting, so it's already calculated like 4-5 times every page load, even before this change. I think if we are to optimize is, it could be calculated once at app startup and then be cached in memory.

Comment (wxiaoguang) [reply templates/base/head_script.tmpl]: Yup, it's better to remove all `MD5 AppVer` but it's a little out of scope so I didn't mention it (I have removed it including the `MD5` helper function in another discarded PR).

For this PR, I think there is no need to introduce more MD5 calculations. Others `MD5` can be removed in this PR together at this time or be done in the future.

Comment (wxiaoguang) [reply templates/base/head_script.tmpl]: ps: I also think using `index.js?v=1.17.1` (`{{AppVer}}`) is clearer than `index.js?v=md5hash`(`{{MD5 AppVer}}`), when users report problems, it's clear about which version they are using.

Comment (silverwind) [reply templates/base/head_script.tmpl]: Version is not optimal in any case, so it should probably be changed to a content hash at some point. Plain version also needs url-escaping I think, so still some processing involved for the escaping.

Comment (silverwind) [reply templates/base/head_script.tmpl]: Also, I imagine escaping a string could be more processing than calculating an md5, so it's really a weak argument.

**Comment by silverwind** [comment]
https://github.com/go-gitea/gitea/pull/20813/commits/8247395de4a53912dc026097d0758bc8f3870a50 introduces `assetVersion` which holds the escaped query that is appended to the html-based and worker assets.

**Comment by silverwind** [comment]
I ended up changing it to a hash again, URLs were to ugly with the plain version string, and I think it reflects badly on us if we don't encode it to a hash.

image

**Comment by wxiaoguang** [comment]
The reason why I prefer to use `index.js?v=1.17.1` is: some users just report frontend errors like "Error occurs on index.js?v=.... line x" from console. Using app version directly will make maintainers to know the version in the first time, without asking more questions, and I do not think it's ugly.

Without introducing more unnecessary `MD5`, either approach LGTM.

File: templates/base/head_script.tmpl
Comment (wxiaoguang) [reply templates/base/head_script.tmpl]: > Version is not optimal in any case, so it should probably be changed to a content hash at some point. Plain version also needs url-escaping I think, so still some processing involved for the escaping.

I do not see why it's not optimal. While content hash also works, either LGTM. The Go template will do the escaping automatically. And JS can also handle the escaping correctly. Even if without escaping, the chars in "AppVersion" are not dangerous, the only ambiguity char plus `+` won't cause any problem even if it is left as it is.

> Also, I imagine escaping a string could be more processing than calculating an md5, so it's really a weak argument.

That's just imagine.

```go
func BenchmarkUrlEscape(b *testing.B) {
	for i := 0; i < b.N; i++ {
		url.QueryEscape("1.17.1")
	}
}
func BenchmarkMd5(b *testing.B) {
	for i := 0; i < b.N; i++ {
		md5.Sum([]byte("1.17.1"))
	}
}
```

```
cpu: Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
BenchmarkUrlEscape
BenchmarkUrlEscape-12    	56735932	        19.01 ns/op
BenchmarkMd5
BenchmarkMd5-12          	10553167	       108.7 ns/op
```

**Comment by silverwind** [comment]
I think it looks ugly and I don't really see a point in this helping bug reports, users are forced to enter a version into a bug report anyways, and it's visible on every page as well.

File: modules/setting/setting.go
Comment (silverwind) [inline modules/setting/setting.go]:
```suggestion
	// AssetVersion holds a opaque value that is used for cache-busting assets
```

**Comment by wxiaoguang** [comment]
> I think it looks ugly and I don't really see a point in this helping bug reports, users are forced to enter a version into a bug report anyways, and it's visible on every page as well.

Because the "forced version" only exists in GitHub issue template, while I collect many user feedbacks from non-github channels, like chats, forums, etc. 😄 Many users just post a screenshot and say "hey, it doesn't work"

**Comment by zeripath** [comment]
I think I have to agree with @wxiaoguang here, what's the benefit of SHAing the version here? You may as well just put the version in.

**Comment by silverwind** [comment]
Consistency I guess, 8 character opaque value used both by backend and webpack. We can change the mechanism at any time later and it will still look consistent. I don't care which hashing mechanism, but I generally avoid deprecated hashes like MD5 and SHA1, that's why I went with SHA256.

**Comment by lunny** [comment]
> The reason why I prefer to use `index.js?v=1.17.1` is: some users just report frontend errors like "Error occurs on index.js?v=.... line x" from console. Using app version directly will make maintainers to know the version in the first time, without asking more questions, and I do not think it's ugly.
> 
> Without introducing more unnecessary `MD5`, either approach LGTM.

But it will make user confusing that which the version mean? It's the index.js' special version or the whole site version?

**Comment by wxiaoguang** [comment]
> But it will make user confusing that which the version mean? It's the index.js' special version or the whole site version?

Does it matter? Is there really any user would read or parse that query parameter? That's only valuable for maintainers and developers.

ps: I do not insist for it as my comment above, either approach LGTM.

**Comment by silverwind** [comment]
It's back at AppVer again

image

**Review by zeripath — APPROVED** [review]

**Comment by lunny** [comment]
> > But it will make user confusing that which the version mean? It's the index.js' special version or the whole site version?
> 
> Does it matter? Is there really any user would read or parse that query parameter? That's only valuable for maintainers and developers.
> 
> ps: I do not insist for it as my comment above, either approach LGTM.

Yes, I mean it may make contributors confusing. We have already over 1 thousand contributors from beginning.

**Comment by wxiaoguang** [comment]
> Yes, I mean it may make contributors confusing. We have already over 1 thousand contributors from beginning.

Can you provide some details about how could they be confused and what's the negative effect for it?

ps: it could make the URL not ugly without double escaping.

**Comment by wxiaoguang** [comment]
I made a commit.  Major changes:

* Remove the `url.QueryEscape`, the Go HTML template will escape it correctly. Otherwise there will be double-escaping.
* Rename the JS `assetVersion` to `assetVersionEncoded` to declare that it is an already-encoded value and can be used in URL directly.
* Replace the plus `+` to a URL-safe char `~` (others could are `.` and `_`), then the URL becomes more readable.



Now, that's what you will see:

```
/assets/css/index.css?v=1.18.0~dev-324-gfa8fc5d3a
/assets/js/index.js?v=1.18.0~dev-324-gfa8fc5d3a
/assets/js/eventsource.sharedworker.js?v=1.18.0~dev-324-gfa8fc5d3
```

File: templates/base/head_script.tmpl
Comment (silverwind) [inline templates/base/head_script.tmpl:12]: Why need to escape manually here? Shouldn't `html/template` escape it, or is that contextual escaping failing?

Comment (wxiaoguang) [reply templates/base/head_script.tmpl:12]: Here is a JS context, `html/template` will only encode the AssetVersion into a JS string.

**Comment by silverwind** [comment]
> Seems like useless extra work to me and I think we're nitpicking too much.

Don't really agree as it mangles the data and other places in JS code might need the value in original form later, but whatever, it works for the current purpose.

File: templates/base/head_script.tmpl
Comment (silverwind) [reply templates/base/head_script.tmpl:12]: I see. [Go's autoescaping](https://go.dev/src/html/template/escape.go) is really an insane way of doing escaping, as it forces the user to know the complex conditions when it will escape. A much more sane way would be how Handlebars did it, e.g. escape everything and let user opt-out via triple quotes, e.g. `{{{value}}}`.

**Comment by wxiaoguang** [comment]
> > Seems like useless extra work to me and I think we're nitpicking too much.
> 
> Don't really agree as it mangles the data and other places in JS code might need the value in original form later, but whatever, it works for the current purpose.

Hmm, I didn't mean to make the problem more complex. My initial thought was that no need to do more `MD5`, and shared some thoughts about the version format when I was trying catch frontend bugs for users. For different version formats, I have written "either LGTM" in all my comments. 😁

**Pull request merged**
